### ISSUE TITLE
Double Free for DynamicData for Sequences of Nested Structs (with minimal example)

### ISSUE THREAD

**Issue by methylDragon** [eProsima/Fast-DDS#2970]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

1. Create a dynamic data consisting of a sequence/array of nested structs
2. Populate that sequence with nested struct elements
3. Exit cleanly

### Current behavior

1. Create a dynamic data consisting of a sequence/array of nested structs
2. Populate that sequence with nested struct elements
3. Exit cleanly <--- Errors print here

Specifically, the following error is printed per sequence/array element created.

```
[DYN_TYPES Error] Error deleting DynamicData. It isn't registered in the factory -> Function delete_data
```

### Steps to reproduce

I've prepared a minimal example:
[methylDragon/fastdds-pr-minimal-examples](https://github.com/methylDragon/fastdds-pr-minimal-examples)

Once built, run:

```
cd build

./src/nested-sequences-bug/could_not_delete
```

The relevant code can be found [here](https://github.com/methylDragon/fastdds-pr-minimal-examples/tree/main/src/nested-sequences-bug/src).

### Fast DDS version/commit

FastDDS 2.8.x: https://github.com/eProsima/Fast-DDS/commit/78473a0de3d878a71565ee5d6c8e1e01c34beb70

The examples workspace is using that specific SHA in the submodule, for convenience.

### Platform/Architecture

Other. Please specify in Additional context section.

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

This is on 22.04, but I don't think the platform matters in this case.

The error is being emitted from [DynamicDataFactory](https://github.com/eProsima/Fast-DDS/blob/master/src/cpp/dynamic-types/DynamicDataFactory.cpp#L205)'s destructor call.

What I suspect is happening is, as each dynamic data is added to the sequence/array, the data gets registered once to the factory on creation. But on destruction, the factory destroys the instance, but the sequence/array destruction triggers a second destruction.

This causes the double free. (I think)

### Relevant log output

```shell
nested_sequence: 
	[0] = <struct/bitset>
		inner_uint32: 1

	[1] = <struct/bitset>
		inner_uint32: 2

2022-09-22 16:04:48.127 [DYN_TYPES Error] Error deleting DynamicData. It isn't registered in the factory -> Function delete_data
2022-09-22 16:04:48.127 [DYN_TYPES Error] Error deleting DynamicData. It isn't registered in the factory -> Function delete_data
```
```


### Network traffic capture

_No response_

**Comment by methylDragon** [comment]
This bug is still happening, are there any updates?

**Comment by methylDragon** [comment]
If it helps, I found that commenting out [this line](https://github.com/eProsima/Fast-DDS/blob/master/src/cpp/dynamic-types/DynamicDataFactory.cpp#L151) fixes the issue. But I don't know what the fallout of removing `default_array_value_` is

**Comment by JLBuenoLopez-eProsima** [comment]
There are several known issues related with Dynamic Types in Fast DDS. Thus, a [refactor is being undertaken in order to fix the Dynamic Types implementation](https://github.com/eProsima/Fast-DDS/pull/3392). It is expected to be released with Fast DDS v2.11.0 scheduled for June 2023.

**Comment by elianalf** [comment]
The [Dynamic Types refactor](https://github.com/eProsima/Fast-DDS/pull/4190) has been integrated in Fast DDS v3.0.0 which will be released soon. This issue may be solved in the new version of Fast DDS, so I will close this issue for now. If the issue persists, feel free to reopen it.
